File: go/receipt_fixtures_test.go
// SPDX-FileCopyrightText: Copyright (C) 2026 provide.io llc
// SPDX-License-Identifier: Apache-2.0
// SPDX-Comment: Part of provide-telemetry.

package telemetry

import (
	"crypto/hmac"
	"crypto/sha256"
	"encoding/hex"
	"os"
	"path/filepath"
	"testing"

	"gopkg.in/yaml.v3"
)

// The canonical governance vectors in spec/receipt_fixtures.yaml were produced
// by independent implementations — rfc8785 for the canonical JSON, Python's
// hmac for the signatures — so reproducing them byte for byte is evidence of
// cross-language agreement rather than of agreeing with ourselves.

type _receiptFixture struct {
	ID            string `yaml:"id"`
	Key           string `yaml:"key"`
	Input         any    `yaml:"input"`
	Normalized    any    `yaml:"normalized"`
	CanonicalJSON string `yaml:"canonical_json"`
	ReceiptID     string `yaml:"receipt_id"`
	Timestamp     string `yaml:"timestamp"`
	FieldPath     string `yaml:"field_path"`
	Action        string `yaml:"action"`
	OriginalHash  string `yaml:"original_hash"`
	Payload       string `yaml:"payload"`
	Signature     string `yaml:"signature"`
}

type _receiptFixtureFile struct {
	Version int               `yaml:"version"`
	Cases   []_receiptFixture `yaml:"cases"`
}

func _loadReceiptFixtures(t *testing.T) []_receiptFixture {
	t.Helper()
	path := filepath.Join("..", "spec", "receipt_fixtures.yaml")
	raw, err := os.ReadFile(path) //nolint:gosec // fixed in-repo fixture path
	if err != nil {
		t.Fatalf("reading %s: %v", path, err)
	}
	var file _receiptFixtureFile
	if err := yaml.Unmarshal(raw, &file); err != nil {
		t.Fatalf("parsing %s: %v", path, err)
	}
	if len(file.Cases) == 0 {
		t.Fatalf("%s declared no cases", path)
	}
	return file.Cases
}

// TestReceiptFixturesCanonicalJSON pins the RFC 8785 serialization, from both
// the value as captured and the value after normalization. The two forms differ
// only where JSON cannot encode the input — NaN and ±Infinity — and both must
// land on the same bytes, which is what makes the digest independent of how a
// non-finite number reached the hook.
func TestReceiptFixturesCanonicalJSON(t *testing.T) {
	for _, tc := range _loadReceiptFixtures(t) {
		t.Run(tc.ID, func(t *testing.T) {
			if got := CanonicalJSON(tc.Input); got != tc.CanonicalJSON {
				t.Errorf("canonical JSON of input:\n got: %s\nwant: %s", got, tc.CanonicalJSON)
			}
			if got := CanonicalJSON(tc.Normalized); got != tc.CanonicalJSON {
				t.Errorf("canonical JSON of normalized:\n got: %s\nwant: %s", got, tc.CanonicalJSON)
			}
		})
	}
}

// TestReceiptFixturesSignedReceipts pins the whole receipt: the hash over the
// canonical bytes, the payload field order, and the HMAC over that payload.
func TestReceiptFixturesSignedReceipts(t *testing.T) {
	for _, tc := range _loadReceiptFixtures(t) {
		t.Run(tc.ID, func(t *testing.T) {
			receipt := signReceipt(tc.Input, receiptFields{
				ReceiptID:   tc.ReceiptID,
				Timestamp:   tc.Timestamp,
				ServiceName: "fixture-svc",
				FieldPath:   tc.FieldPath,
				Action:      tc.Action,
			}, tc.Key)

			if receipt.OriginalHash != tc.OriginalHash {
				t.Errorf("original_hash:\n got: %s\nwant: %s", receipt.OriginalHash, tc.OriginalHash)
			}
			if got := receiptPayload(receipt); got != tc.Payload {
				t.Errorf("payload:\n got: %s\nwant: %s", got, tc.Payload)
			}
			if receipt.HMAC != tc.Signature {
				t.Errorf("signature:\n got: %s\nwant: %s", receipt.HMAC, tc.Signature)
			}
		})
	}
}

// TestReceiptFixturesHashCoversCanonicalBytes re-derives each digest from the
// fixture's own canonical_json string, so a bug that happened to affect both
// our serializer and our hasher the same way cannot hide.
func TestReceiptFixturesHashCoversCanonicalBytes(t *testing.T) {
	for _, tc := range _loadReceiptFixtures(t) {
		t.Run(tc.ID, func(t *testing.T) {
			sum := sha256.Sum256([]byte(tc.CanonicalJSON))
			if got := hex.EncodeToString(sum[:]); got != tc.OriginalHash {
				t.Errorf("fixture is self-inconsistent: %s vs %s", got, tc.OriginalHash)
			}
			mac := hmac.New(sha256.New, []byte(tc.Key))
			mac.Write([]byte(tc.Payload)) //nolint:errcheck // hash.Hash.Write never errors
			if got := hex.EncodeToString(mac.Sum(nil)); got != tc.Signature {
				t.Errorf("fixture signature is self-inconsistent: %s vs %s", got, tc.Signature)
			}
		})
	}
}

File: typescript/tests/receipt-fixtures.test.ts
// SPDX-FileCopyrightText: Copyright (c) 2025-2026 provide.io llc. All rights reserved.
// SPDX-License-Identifier: Apache-2.0
//
// Cross-language governance vectors: spec/receipt_fixtures.yaml.
//
// These are the only receipt tests that can fail for being *different from the
// other SDKs* rather than different from themselves. The canonical strings and
// digests were produced by independent implementations — `rfc8785` for JCS,
// Python's `hmac` for the signature — so reproducing them byte for byte is
// evidence of agreement, not of self-consistency.

import { readFileSync } from 'node:fs';
import { describe, expect, it } from 'vitest';
import { parse } from 'yaml';
import { specFixturePath } from './support/spec-fixtures.js';
import { hmacSha256Hex, sha256Hex } from '../src/hash.js';
import { canonicalJson, receiptPayload, signReceipt } from '../src/receipts.js';

interface ReceiptVector {
  id: string;
  key: string;
  input: unknown;
  normalized: unknown;
  canonical_json: string;
  receipt_id: string;
  timestamp: string;
  field_path: string;
  action: string;
  original_hash: string;
  payload: string;
  signature: string;
}

const FIXTURES = specFixturePath('receipt_fixtures.yaml');
const vectors = (parse(readFileSync(FIXTURES, 'utf8')) as { cases: ReceiptVector[] }).cases;
const enc = new TextEncoder();

describe('receipt vectors', () => {
  it('loads every committed case', () => {
    // Guards against a silently empty suite: a parse that yields no cases would
    // otherwise make every it.each below vacuously pass.
    expect(vectors.length).toBeGreaterThanOrEqual(7);
  });

  it.each(vectors)('canonicalizes $id to the committed JCS bytes', (vector) => {
    expect(canonicalJson(vector.normalized)).toBe(vector.canonical_json);
  });

  it.each(vectors)('signs $id to the committed payload and signature', (vector) => {
    const receipt = signReceipt(vector.normalized, {
      receiptId: vector.receipt_id,
      timestamp: vector.timestamp,
      fieldPath: vector.field_path,
      action: vector.action,
      key: enc.encode(vector.key),
    });
    expect(receipt.originalHash).toBe(vector.original_hash);
    expect(receiptPayload(receipt)).toBe(vector.payload);
    expect(receipt.hmac).toBe(vector.signature);
  });

  it.each(vectors)('derives $id digests from the canonical bytes, not the object', (vector) => {
    expect(sha256Hex(vector.canonical_json)).toBe(vector.original_hash);
    expect(hmacSha256Hex(enc.encode(vector.key), enc.encode(vector.payload))).toBe(
      vector.signature,
    );
  });
});

describe('non-finite normalization', () => {
  it('canonicalizes the raw input to the same bytes as the normalized form', () => {
    // The fixture carries `input` (with NaN/±Infinity) and `normalized` (with
    // nulls) separately so normalization is part of the contract rather than
    // left to each SDK to invent. YAML cannot express NaN portably, so the raw
    // values are rebuilt here from the case that declares them.
    const vector = vectors.find((v) => v.id === 'non_finite_normalization');
    expect(vector).toBeDefined();
    const raw = {
      ratio: Number.NaN,
      ceiling: Number.POSITIVE_INFINITY,
      floor: Number.NEGATIVE_INFINITY,
      ok: 2.0,
    };
    expect(canonicalJson(raw)).toBe((vector as ReceiptVector).canonical_json);
  });
});

describe('signReceipt without a key', () => {
  it('leaves the signature empty rather than signing under an empty key', () => {
    const vector = vectors[0] as ReceiptVector;
    const receipt = signReceipt(vector.normalized, {
      receiptId: vector.receipt_id,
      timestamp: vector.timestamp,
      fieldPath: vector.field_path,
      action: vector.action,
    });
    expect(receipt.originalHash).toBe(vector.original_hash);
    expect(receipt.hmac).toBe('');
    expect(receipt.serviceName).toBe('unknown');
  });

  it('carries the configured service name onto the receipt', () => {
    const vector = vectors[0] as ReceiptVector;
    const receipt = signReceipt(vector.normalized, {
      receiptId: vector.receipt_id,
      timestamp: vector.timestamp,
      fieldPath: vector.field_path,
      action: vector.action,
      serviceName: 'billing',
    });
    expect(receipt.serviceName).toBe('billing');
  });
});
